src/server/utils/auth.js

```diff
@@ -130,46 +136,86 @@ async function _verifyToken(token, secretOrPublicKey, options = {}, validScopes,
-		if (config.server && config.server.mode === 'public') {
-			return next();
-		} else {
 			// get bearer token
-			const bearerToken = _parseBearerToken(req);
+			const bearerToken = parseBearerToken(req);
 
 			if (bearerToken) {
-				logger.debug('Found bearer token in request');
+				// decode token
 				const decodedToken = jwt.decode(bearerToken, { complete: true });
+				if (decodedToken) {
+
+					let validToken;
+
+					// get client
+					let client = await service.getClient(decodedToken.aud);
+					if (client) {
+						// verify token and signature
+						let [error, token] = await handler(verifyToken(decodedToken, bearerToken, client, config.auth.resourceServer));
+						if (error) {
+							logger.error(error);
+							return next(errors.unauthorized());
+						}
+						validToken = token;
+
+					} else {
+						// if client isn't found, attempt to introspect
+						if (config.auth.discoveryUrl) {
+
+							let [ introspectionError, introspection ] = await handler(introspect(bearerToken, config));
+							if (introspectionError) {
+								logger.error(introspectionError);
+								return next(errors.unauthorized());
+							}
+							validToken = introspection;
+						}
+					}
 
-				if (decodedToken && decodedToken.header && decodedToken.header.alg) {
-					if (decodedToken.header.alg.startsWith('HS')) {
-						logger.debug('HS Algorithm');
-						// IF HS*** algorith, validate signature based on secret key
-						_verifyToken(bearerToken, config.auth.secret, {}, validScopes, config.auth, next);
-
-					} else if (decodedToken.header.alg.startsWith('RS')) {
-						logger.debug('RHS Algorithm');
 
-						// IF RS*** algorithm, validate signature based on certificate
-						// Get public key (update this to point to the correct public key path)
-						const jwtKey = _.find(config.jwkSet.keys, ['kid', decodedToken.header.kid]);
-						const pem = jwkToPem(jwtKey);
-						_verifyToken(bearerToken, pem, {algorithms: [decodedToken.header.alg]}, validScopes, config.auth, next);
+					// we have a valid token at this point, let's verify the scope
+					let scopes;
+					// if scopes are on the token
+					if (validToken.scope) {
+						scopes = getValidScopes(validToken.scope, allowedScopes);
+					} else if (client.scope) {
+						// let's look at the scopes of the client
+						scopes = getValidScopes(client.scope, allowedScopes);
+					} 
+
+					// verify if token has one of the required scopes
+					if (scopes && scopes.length > 0) {
+						// attach patient id to req or validate request and patient
+						// TODO: verify this as it could be done in different ways
+						// if (validToken.context && validToken.context.patient) {
+						// 	req.patientId = validToken.context.patient;
+						// }
+
+						// req.patient is a signed or hash that the resource server knows
+						// how to validate to get the real patient id
+
+						// validation complete
+						return next();
+					} else {
+						return next(errors.unauthorized());
```

Comment:
We should probably log issues such as authentication failures and scope mismatches, for auditing purposes, as long as the log messages don't contain any secrets